types/react/index.d.ts

```diff
@@ -90,6 +90,42 @@ declare namespace React {
     type RefCallback<T> = { bivarianceHack(instance: T | null): void }["bivarianceHack"];
     type Ref<T> = RefCallback<T> | RefObject<T> | null;
     type LegacyRef<T> = string | Ref<T>;
+    /**
+     * Gets the instance type for a React element. The instance will be different for various component types:
+     *
+     * - React class components will be the class instance. So if you had `class Foo extends React.Component<{}> {}`
+     *   and used `React.ElementRef<typeof Foo>` then the type would be the instance of `Foo`.
+     * - React stateless functional components do not have a backing instance and so `React.ElementRef<typeof Bar>`
+     *   (when `Bar` is `function Bar() {}`) will give you the `undefined` type.
+     * - JSX intrinsics like `div` will give you their DOM instance. For `React.ElementRef<'div'>` that would be
+     *   `HTMLDivElement`. For `React.ElementRef<'input'>` that would be `HTMLInputElement`.
+     * - React stateless functional components that forward a `ref` will give you the `ElementRef` of the forwarded
+     *   to component.
+     *
+     * `C` must be the type _of_ a React component so you need to use typeof as in React.ElementRef<typeof MyComponent>.
+     *
+     * @todo In Flow, this works a little different with forwarded refs and the `AbstractComponent` that
+     *       `React.forwardRef()` returns.
+     */
+    type ElementRef<
+        C extends
+            | { new (props: any): Component<any> }
+            | (() => ReactElement | null)
+            | keyof JSX.IntrinsicElements
+            | ForwardRefExoticComponent<any>
+    > = C extends { new (props: any): Component<any> }
+        ? InstanceType<C>
+        : C extends (() => ReactElement | null)
+        ? undefined
+        : C extends keyof JSX.IntrinsicElements
+        ? JSX.IntrinsicElements[C] extends DOMAttributes<infer E>
+            ? E
+            : never
+        : C extends ForwardRefExoticComponent<infer FP>
```

Comment:
Does this actually work? I had thought that since this is lower in the conditional type chain, it wouldn't be reached.

Exotic components are technically function components (as far as the types are concerned), so shouldn't this be inside the `true` clause of the FP check?

```ts
type ElementRef<C> = C extends ((props: any) => ReactElement | null)
  ? C extends ForwardRefExoticComponent<infer P>
    ? P extends RefAttributes<infer T> ? T : never
  : undefined
```

Reply:
After fixing the problem of not matching function components with props, this indeed started failing, so updated the order of precedence to fix this 👍